graphql_compiler/query_pagination/pagination_planning.py

```diff
@@ -1,53 +1,172 @@
+    desired_resolution: int
+    message: str = field(init=False)
+
+    def __post_init__(self):
+        """Initialize a human-readable message."""
+        self.message = (
+            "Pagination would have been more successful if more quantiles were provided "
+            "for {}.{}. Currently there are {}, ideally there should be {}".format(
+                self.vertex_name, self.field_name, self.current_resolution, self.desired_resolution
+            )
+        )
+
 
+class VertexPartition(NamedTuple):
+    """Plan to split the query at a certain vertex into a certain number of pages."""
 
-# The intent to split the query at a certain vertex into a certain number of pages.
-VertexPartition = namedtuple(
-    "VertexPartition",
-    (
-        "query_path",  # Tuple[field name : str] leading to the vertex to be split
-        "number_of_splits",  # The number of subdivisions intended for this vertex
-    ),
-)
+    # field names leading to the vertex to be split
+    query_path: Tuple[str, ...]
 
+    # field to use for pagination
+    pagination_field: str
 
-# The intent to split the query with a combination of VertexPartitions
-PaginationPlan = namedtuple("PaginationPlan", ("vertex_partitions",))  # List[VertexPartition]
+    # The number of subdivisions intended for this vertex
+    number_of_splits: int
 
 
-# TODO(bojanserafimov): Make this function return a best effort pagination plan
-#                       when a good one is not found instead of returning None.
-def try_get_pagination_plan(schema_info, query_ast, number_of_pages):
-    """Make a PaginationPlan for the given query and number of desired pages if possible."""
+class PaginationPlan(NamedTuple):
+    """Plan to split the query with a combination of VertexPartitions."""
+
+    vertex_partitions: Tuple[VertexPartition, ...]
+
+
+def get_num_pages_generated_by_plan(plan: PaginationPlan) -> int:
+    """Return the number of pages that a PaginationPlan would generate."""
+    number_of_pages = 0
+    for vertex_partition in plan.vertex_partitions:
+        number_of_pages *= vertex_partition.number_of_splits
+    return number_of_pages
+
+
+def get_pagination_plan(
+    schema_info: QueryPlanningSchemaInfo, query_ast: DocumentNode, number_of_pages: int
+) -> Tuple[PaginationPlan, List[PaginationAdvisory]]:
+    """Make a best-effort PaginationPlan and advise on how to improve statistics.
+
+    Might paginate to fewer than the desired number of pages if no good pagination plan
+    is found. This can happen when there's not enough data, or when the planner is not
+    smart enough to find a good plan. In that case it will return along with the result
+    a list of PaginationAdvisorys that indicate why the pagination was not successful.
+    """
     definition_ast = get_only_query_definition(query_ast, GraphQLError)
 
+    if number_of_pages <= 0:
+        raise AssertionError(
+            u"The number of pages should be at least 1: {}".format(number_of_pages)
```

Comment:
To be more clear about what happened, maybe include `at least 1, but {} were requested.`?